sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala

```diff
@@ -69,6 +80,21 @@ object CharVarcharUtils {
     case _ => dt
   }
 
+  /**
+   * Replaces CharType/VarcharType with StringType recursively in the given data type, with a
+   * warning message if it has char or varchar types
+   */
+  def logWarningAndReplaceCharVarcharWithString(dt: DataType): DataType = {
+    if (hasCharVarchar(dt)) {
+      logWarning("In Spark 3.0 and earlier, char/varchar is as same as string type, since" +
```

Comment:
`The Spark cast operator does not support char/varchar type and simply treat them as string type. Please use string type directly to avoid confusion.`